n3fit/src/n3fit/backends/keras_backend/multi_dense.py

```diff
@@ -0,0 +1,191 @@
+    >>>     Sequential([
+    >>>         Dense(units=8, kernel_initializer=GlorotUniform(seed=42 + r)),
+    >>>         Dense(units=4, kernel_initializer=GlorotUniform(seed=52 + r)),
+    >>>         ])
+    >>>     for r in range(replicas)
+    >>>     ]
+    >>> gridsize, features = 100, 2
+    >>> multi_dense_model.build(input_shape=(None, gridsize, features))
+    >>> for single_model in single_models:
+    >>>     single_model.build(input_shape=(None, gridsize, features))
+    >>> test_input = tf.random.uniform(shape=(1, gridsize, features))
+    >>> multi_dense_output = multi_dense_model(test_input)
+    >>> single_dense_output = tf.stack([single_model(test_input) for single_model in single_models], axis=1)
+    >>> tf.reduce_all(tf.equal(multi_dense_output, single_dense_output))
+
+    Parameters
+    ----------
+    replica_seeds: List[int]
+        List of seeds per replica for the kernel initializer.
+    kernel_initializer: Initializer
+        Initializer class for the kernel.
+    replica_input: bool (default: True)
+        Whether the input already contains multiple replicas.
+    """
+
+    def __init__(
+        self,
+        replica_seeds: List[int],
+        kernel_initializer: Initializer,
+        replica_input: bool = True,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.replicas = len(replica_seeds)
+        self.replica_seeds = replica_seeds
+        self.kernel_initializer = MultiInitializer(
+            single_initializer=kernel_initializer, replica_seeds=replica_seeds
+        )
+        self.bias_initializer = MultiInitializer(
+            single_initializer=self.bias_initializer, replica_seeds=replica_seeds
+        )
+        self.replica_input = replica_input
+
+        self.matmul = None
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.kernel = self.add_weight(
+            name="kernel",
+            shape=(self.replicas, input_dim, self.units),
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+        )
+        if self.use_bias:
+            self.bias = self.add_weight(
+                name="bias",
+                shape=(self.replicas, 1, self.units),
+                initializer=self.bias_initializer,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+            )
+        else:
+            self.bias = None
+        self.input_spec.axes = {-1: input_dim}
+        self.built = True
+
+        if self.replicas == 1:
+            if self.replica_input:
+                self.matmul = lambda inputs: tf.tensordot(inputs, self.kernel[0], [[-1], [0]])
+            else:
+                # Manually add replica dimension
+                self.matmul = lambda inputs: tf.expand_dims(
+                    tf.tensordot(inputs, self.kernel[0], [[-1], [0]]), axis=1
+                )
```

Comment:
```suggestion
            # Using tensordot here for numerical stability with 4.0 fits
            # TODO: benchmark against the replica-agnostic einsum below and make that default
            if self.replica_input:
                axis = 1
            else:
                axis = 0
            self.matmul = lambda inputs: tf.tensordot(inputs, self.kernel[0], [[-1], [0]], axis=axis)
```
Remind me, why can `self.replica_input` be `True` or `False`?

Reply:
> why can self.replica_input be True or False? 

First layers shares the same inputs for all replicas, in which case it's `False`. Only the later layers have an additional index for the replicas

Reply:
But then in the 1 replica case it doesn't matter.

Reply:
I think it's just for convenience of the output shape

Reply:
> Also, remind me, why can self.replica_input be True or False? (and will that happen in the 1 replica case?)

As Roy said, it's because the input to the first layer is just the x's, which doesn't have a replica dimension yet. (That's independent of the number of replicas being 1 or more, in both cases there is no extra dimension to the first layer's input, but there is to all following layers).

I agree it's not very well named, perhaps I can replace it with just `is_first_layer`?

> Can't you do something like a reshape with the first dimension being (1,) which would work in both cases?

It would be something like this:
```python
            output_shape = input_shape[:-1] + (self.units,)
            if self.replica_input:  # Have to manually add the replica dimension
                output_shape = (1,) + output_shape

            self.matmul = lambda inputs: tf.reshape(
                    tf.tensordot(inputs, self.kernel[0], [[-1], [0]]), output_shape
            )
```
avoiding the if statement, but this doesn't work as the batch dimension is `None` when this is called. It can be set manually of course, but then I think it starts to get less clear than what we started with.

Reply:
I'm quite confused, why in this case you don't need axis?

Reply:
I'm quite confused by your question, what are you referring to exactly? ;P

Maybe this answers your question: note the `self.kernel[0]`, that removes the replica axis from the kernel. (In a failed attempt to reproduce the numerics exactly). I'll see what happens if I remove that, but I don't think the axis will end up in the right place.

Reply:
The original code was:

```python
           if self.replica_input:
                self.matmul = lambda inputs: tf.tensordot(inputs, self.kernel[0], [[-1], [0]])
            else:
                # Manually add replica dimension
                self.matmul = lambda inputs: tf.expand_dims(
                    tf.tensordot(inputs, self.kernel[0], [[-1], [0]]), axis=1
                )
```

The tensordot call in your new example is the same but now you don't seem to need the axis argument while before you did.

Reply:
No the `axis=` is an argument to `tf.expand_dims`, which in the (broken) code above is replaced by `tf.reshape` with a predetermined shape.

Reply:
ah!!! Then we were talking cross-purposes, I missed, the `)` 

That makes it much easier. Let me change my suggestion.

```suggestion
            self.matmul =  lambda inputs: tf.tensordot(inputs, self.kernel[0], [[-1], [0]])
            if not self.replica_input:
                # Manually add replica dimension
                self.matmul = lambda x: tf.expand_dims(self.matmul(x), axis=1))
```
or some flavour thereof.

My main concern was to make clear that it was the same operation happening in both cases, Having an extra `expan_dims` afterwards I'm less worried about. I thought the `tensordot` was different in the two cases, sorry.

Reply:
Yeah the way black formats this is not the clearest.

Ah yes that looks good, just had to modify it slightly to prevent it from going into an infinite recursion.

Reply:
I've rewritten `replica_input` to (`not`) `is_first_layer`, I think it's much clearer like that.